### PR TITLE
WIP for base class TestResult

### PR DESCRIPTION
**Pull request by N-Wouda**
Following #4994, which references to i.a. #1182, I decided to open a new PR for this issue. While #1182 aspires to mostly the same thing, it appears largely stale. I intend to use this base class in multiple places, and have several locations inherit from, or directly use this base class.

After that is set-up, a next step will be to consolidate the children classes and negotiate a proper protocol per module (e.g., the `tsa` tests might need a different parent class than do most `regression` tests and the like).

@josef-pkt input is much appreciated! :)

### CODE REVIEW

**Comment by josef-pkt** [comment]
I think a more likely location is statsmodels.stats.base
Generally this looks fine.

We need some optional attributes that can be listed by default after the main attributes, e.g. attributes in my latest `Holder` https://github.com/statsmodels/statsmodels/pull/4829/files#diff-36a55e208f6d1db357321ee87ef3ab77R1385 

Also in most cases we don't specify critical values but report p-values and include the distribution, e.g. normal, t, F, chi2.  In those cases we can leave critical values at None and don't print them.
Examples associated with model test results are in stats.contrast (but t-test results do a lot more than what standard test results show, e.g. computing confint)

File: statsmodels/stats/base.py [hunk 1]
```diff
@@ -99,3 +102,97 @@ def summary(self):
...
+
+    def __init__(self, test_name, statistics, **kwargs):
+        self.test_name = test_name
+        self.statistics = statistics
+
+        for key, value in kwargs.items():
+            if key in TestResult._options:
+                setattr(self, key, value)
+
+    def __getattr__(self, item):
```
Comment (N-Wouda) [inline statsmodels/stats/base.py]: My thinking here is to preserve interface compatibility between old-style test results (e.g. `Bunch`, or `ResStore`) and new-style tests where possible. I intend `Statistics` to be a new-style store object, against which the `TestResult` will always check if the attribute cannot be found on itself.

If it is found in the statistics attribute, the item there is returned and a `DepreciationWarning` is raised explaining the new behaviour. This way we can start using the new-style test results without breaking too much of existing, external code.

**Comment by jbrockmendel** [comment]
It might be worth looking at how arch does this.  A lot of thought went into that.

**Comment by N-Wouda** [comment]
>It might be worth looking at how arch does this. A lot of thought went into that.

I'll have a look at that later today.

**Comment by N-Wouda** [comment]
@josef-pkt I agree, but before then I'd like to continue working on it for a bit - there's some additional ideas I want to test, but I have not yet had the time to implement (including the optional unpacking for backwards compatibility that I'd like to test some more and implement in more places).

I should be able to make the time during my upcoming Christmas break.

**Comment by N-Wouda** [comment]
I recently realised I never did quite finish this. I have some time over the summer, and hope to continue this then (might refactor some of this as I have a couple of new ideas on how to proceed).

One thing I'd like to do is make the return of a test object non-optional. Code that currently depends on the return of a tuple may still be supported by having the test object unpack itself if requested, by say making the test object inherit from `abc.Sequence`. Upon such request we can then raise a deprecation warning that these tests now return proper test objects, and code should be updated to support this.

**Comment by josef-pkt** [comment]
FYI: the default class for hypothesis test and similar returns is now a `HolderTuple` class or a subclass of it with additional methods
It has mainly tuple unpacking and repr.

**Pull request closed without merging**